Sprint 15

SnakeAssassin opened the pull request:
лизован слой работы с базой данных с использованием абстракций:
добавлены классы TrackerStore, TrackerCategoryStore, TrackerRecordStore для абстрагирования Core Data;
добавлены Entity для хранения Tracker, TrackerCategory, TrackerRecord;
у каждой Entity есть суффикс, который обозначает, что она используется для хранения в Core Data (например, TrackerCoreData);
в модели (файле с расширением *.xcdatamodeld) указаны взаимосвязи между Entity;
создан один Persistent Container, который управляет тремя типами Store;
использован NSFetchedResultsController для получения обновлений от Core Data;
можно создавать привычки и нерегулярные события — и после перезапуска приложения информация отображается; слой хранения данных на базе Core Data выполняет задачи;
приложение надёжно абстрагировано от Core Data: строчка import CoreData есть только в файлах Store.swift и месте создания NSPersistentContainer.

Commit:
Реализован слой работы с базой данных с использованием абстракций:
добавлены классы TrackerStore, TrackerCategoryStore, TrackerRecordStore для абстрагирования Core Data;
добавлены Entity для хранения Tracker, TrackerCategory, TrackerRecord;
у каждой Entity есть суффикс, который обозначает, что она используется для хранения в Core Data (например, TrackerCoreData);
в модели (файле с расширением *.xcdatamodeld) указаны взаимосвязи между Entity;
создан один Persistent Container, который управляет тремя типами Store;
использован NSFetchedResultsController для получения обновлений от Core Data;
можно создавать привычки и нерегулярные события — и после перезапуска приложения информация отображается; слой хранения данных на базе Core Data выполняет задачи;
приложение надёжно абстрагировано от Core Data: строчка import CoreData есть только в файлах Store.swift и месте создания NSPersistentContainer.

Commit:
Реализована CoreData

helecta commented on Tracker/Stores/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,142 @@
+    private var updatedIndexes: IndexSet?
+    
+    private let uiColorMarshalling = UIColorMarshalling()
+    weak var delegate: TrackerCategoryStoreDelegate?
```
Не забывай организовывать код следующим образом:

- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы
- IBAction

Подробнее почитать можно тут: [Организация кода](https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization)

helecta commented on Tracker/Stores/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,142 @@
+    
+    // MARK: Initialization
+    convenience override init() {
+        let context = (UIApplication.shared.delegate as! AppDelegate).persistentContainer.viewContext
```
Тут force cast: `UIApplication.shared.delegate as! AppDelegate`, стоит переписать без него

helecta commented on Tracker/Stores/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,142 @@
+    // MARK: Initialization
+    convenience override init() {
+        let context = (UIApplication.shared.delegate as! AppDelegate).persistentContainer.viewContext
+        try! self.init(context: context)
```
тоже force операция, стоит убрать

helecta commented on Tracker/Stores/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,142 @@
+    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange anObject: Any, at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?) {
+        switch type {
+        case .delete:
+            if let indexPath = indexPath {
```
лучше вынести `let indexPath = indexPath` перед `switch`, чтобы каждый раз внутри `case` не проверять. А еще лучше использовать `guard let indexPath = indexPath`

helecta commented on Tracker/Stores/TrackerRecordStore.swift:
```diff
@@ -0,0 +1,69 @@
+    // MARK: Initialization
+    convenience override init() {
+        let context = (UIApplication.shared.delegate as! AppDelegate).persistentContainer.viewContext
+        try! self.init(context: context)
```
force операции, такие же как в TrackerCategoryStore

helecta commented on Tracker/Stores/TrackerStore.swift:
```diff
@@ -0,0 +1,84 @@
+    // Удобный инициализатор
+    convenience override init() {
+        guard let appDelegate = UIApplication.shared.delegate as? AppDelegate else {
+            try! self.init(context: nil)
```
force операции

Commit:
Исправлены замечания от ревьюера

Commit:
Исправлены некритичные замечания:
1. добавлена заглушка с текстом: "Что будем отслеживать".
2. Исправлен баг с расписанием "Если после выбора расписания возвращаться назад, но при этом ничего не делать, расписание сбивается"
3. Заменен "Цвет фона для выбранного эмоджи" согласно макету
